File: shared_resources/src/main/java/com/thewizrd/shared_resources/utils/ZonedDateTimeConverter.java
package com.thewizrd.shared_resources.utils;

import com.google.gson.TypeAdapter;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;

import java.io.IOException;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;

public class ZonedDateTimeConverter extends TypeAdapter<ZonedDateTime> {
    private static final DateTimeFormatter formatter = DateTimeUtils.getZonedDateTimeFormatter();

    @Override
    public ZonedDateTime read(JsonReader in) throws IOException {
        return ZonedDateTime.parse(in.nextString(), formatter);
    }

    @Override
    public void write(JsonWriter out, ZonedDateTime value) throws IOException {
        out.value(formatter.format(value));
    }
}

File: shared_resources/src/main/java/com/thewizrd/shared_resources/json/ZonedDateTimeConverter.kt
package com.thewizrd.shared_resources.json

import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import com.thewizrd.shared_resources.utils.DateTimeUtils
import java.io.IOException
import java.time.ZonedDateTime

class ZonedDateTimeConverter : JsonAdapter<ZonedDateTime>() {
    companion object {
        private val formatter = DateTimeUtils.getZonedDateTimeFormatter()
    }

    @Synchronized
    @Throws(IOException::class)
    override fun fromJson(reader: JsonReader): ZonedDateTime? {
        if (reader.peek() == JsonReader.Token.NULL) {
            return reader.nextNull()
        }

        return ZonedDateTime.parse(reader.nextString(), formatter)
    }

    @Synchronized
    @Throws(IOException::class)
    override fun toJson(writer: JsonWriter, value: ZonedDateTime?) {
        if (value == null) {
            writer.nullValue()
        } else {
            writer.value(formatter.format(value))
        }
    }
}